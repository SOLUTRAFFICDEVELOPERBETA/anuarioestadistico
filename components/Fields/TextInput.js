import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import {
    PARAGRAPH,
    QUOTES,
    TITLE,
    SUBTITLE,
    GRID_SIZES,
    CUSTOM_TEXT
} from '../../constants/documents';
// import EditorContext from '../../contexts/editor'
import {
    Box,
    Grid,
    IconButton,
    makeStyles,
    MenuItem,
    Popover,
    TextField,
    Tooltip
} from '@material-ui/core';
import { DeleteForever, DragHandle, FormatColorText } from '@material-ui/icons';
import { DragPreviewImage, useDrag } from 'react-dnd';
import clsx from 'clsx';
import { grey } from '@material-ui/core/colors';
import ColorPick from '../ColorPick';
import EmojiPicker from '../EmojiPicker';

const TEXT_TYPES = {
    [TITLE]: 'Titulo',
    [SUBTITLE]: 'Subtitulo',
    [PARAGRAPH]: 'Párrafo',
    [QUOTES]: 'Cuotas',
    [CUSTOM_TEXT]: 'Personalizado'
};

// Contenedor del Input
const InputContainer = styled.div`
  width: 100%;
  padding: 16px;
  transition: background-color 0.5s ease, z-index 0.5s ease, box-shadow 0.5s ease;
  display: grid;
  grid-template-columns: 0px auto;

  :focus-within {
    background-color: #eee;
    z-index: 5;
    -webkit-box-shadow: 0px 0px 5px 0px rgba(0,0,0,0.75);
    -moz-box-shadow: 0px 0px 5px 0px rgba(0,0,0,0.75);
    box-shadow: 0px 0px 5px 0px rgba(0,0,0,0.75);
    grid-template-columns: 0px auto;
    transition: 0.5s;
  }

  :hover {
    grid-template-columns: 30px auto;
  }

  .dragger {
    align-self: center;
    justify-self: center;
    width: 24px;
    height: 24px;
    visibility: hidden;
    opacity: 0;
    transition: visibility 0s, opacity 0.5s linear;
  }

  :hover .dragger {
    visibility: visible;
    opacity: 1;
  }

  .dragger:hover {
    cursor: move;
  }

  .text-field {
    color: ${(props) => props.color};
    text-align: ${(props) => props.align};
    border: 0px;
    width: 100%;
    padding: 0.5rem;
    max-height: fit-content;
    display: block;
    width: 100%;
    overflow: hidden;
    resize: none;
    min-height: 40px;
    transition: font-size 0.5s ease, color 0.5s ease, text-align 0.5s ease;
  }

  .text-field:hover {
    cursor: pointer;
  }

  .text-field:focus {
    cursor: text;
    background-color: #fff;
  }

  .text-field[contentEditable]:empty::before {
    content: "Haz click para editar";
    color: gray;
  }

  .text-field[data-type=${TITLE}] {
    font-family: ""Roboto", "Helvetica", "Arial", sans-serif";
    font-weight: 400;
    font-size: 2.125rem;
    line-height: 1.235;
    letter-spacing: 0.00735em;
    white-space: normal;
  }

  .text-field[data-type=${SUBTITLE}] {
    font-family: ""Roboto", "Helvetica", "Arial", sans-serif";
    font-weight: 500;
    font-size: 1.25rem;
    line-height: 1.6;
    letter-spacing: 0.0075em;
    white-space: normal;
  }

  .text-field[data-type=${PARAGRAPH}] {
    font-family: ""Roboto", "Helvetica", "Arial", sans-serif";
    font-weight: 400;
    font-size: 1rem;
    line-height: 1.5;
    letter-spacing: 0.00938em;
    white-space: pre-wrap;
  }

  .text-field[data-type=${CUSTOM_TEXT}] {
    font-family: ""Roboto", "Helvetica", "Arial", sans-serif";
    font-weight: 400;
    font-size: ${(props) => props.fontSize}px;
    line-height: 1.5;
    letter-spacing: 0.00938em;
    white-space: pre-wrap;
  }

  .text-field[data-type=${QUOTES}] {
    font-family: ""Roboto", "Helvetica", "Arial", sans-serif";
    font-weight: 400;
    font-size: 1rem;
    line-height: 1.5;
    letter-spacing: 0.00938em;
    white-space: pre-wrap;
  }

  .text-field[data-type=${QUOTES}]::before{
    content: open-quote;
    font-weight: bold;
  }

  .text-field[data-type=${QUOTES}]::after{
    content: close-quote;
    font-weight: bold;
  }
`;

// Estilos del componente
const useStyles = makeStyles((theme) => ({
    option: {
        minWidth: theme.spacing(20)
    },
    dragging: {
        backgroundColor: '#eee',
        // z-index: 5,
        boxShadow: '0px 0px 5px 0px rgba(0,0,0,0.75)'
    }
}));

/**
 * @description Componente para mostrar un campo de texto
 * @param {String} id
 * @param {any} value
 * @param {String} type
 * @param {String} color
 * @param {Number} size
 * @param {String} align
 * @param {String} fontSize
 * @param {Boolean} child
 * @param {Boolean} disableGrid
 * @param {func:() => void} onChange
 * @param {func:() => void} onDelete
 * Consulte los elementos utilizados en {@link https://material-ui.com/ | Material-ui}
 * consulte los elementos DragPreviewImage, useDrag en {@link https://react-dnd.github.io/react-dnd/docs/overview | React DnD Drag and Drop for React}
 */
const TextInput = ({
    id,
    value,
    type,
    color,
    size,
    align = 'start',
    fontSize = 14,
    child,
    disableGrid,
    onChange,
    onDelete
}) => {
    const classes = useStyles();
    const [anchorEl, setAnchorEl] = React.useState(null);
    const [{ isDragging }, drag, preview] = useDrag({
        item: { type: child ? 'CHILD' : 'ITEM', id },
        collect: (monitor) => ({
            isDragging: !!monitor.isDragging()
        })
    });
    // const { onChangeField, onDeleteField } = React.useContext(EditorContext)

    /**
     * Método para abrir el popover
     * @param {Event} event Evento de click de la celda
     */
    const handleClick = (event) => {
        event.preventDefault();
        setAnchorEl(event.currentTarget);
    };

    // Método para cerrar el popover
    const handleClose = () => {
        setAnchorEl(null);
    };

    /**
     * Método para cambiar el contenido del componente
     * @param {Event} event Evento de cambio del campo
     */
    const handleChange = (event) => {
        const {
            currentTarget: { innerText }
        } = event;

        onChange({ value: innerText });
    };

    const handlePushEmoji = ({ emoji }) => onChange({ value: `${value}${emoji}` });

    /**
     * Método para cambiar una propiedad del componente
     * @param {Event} event Evento de cambio del Input
     */
    const handleChangeProps = ({ target: { name, value } }) => {
        onChange({ [name]: value });
    };

    const open = Boolean(anchorEl);

    return (
        <React.Fragment>
            <DragPreviewImage connect={preview} src="/static/img/DragPreview.png" />
            <InputContainer
                ref={drag}
                id={id}
                color={color}
                align={align}
                fontSize={fontSize}
                onContextMenu={handleClick}
                className={clsx({
                    [classes.dragging]: isDragging
                })}>
                <span className="dragger">
                    <DragHandle style={{ fill: grey[500] }} />
                </span>
                <span
                    data-type={type}
                    contentEditable
                    className="text-field"
                    onBlur={handleChange}
                    dangerouslySetInnerHTML={{ __html: value }}
                />
            </InputContainer>
            {open && (
                <Popover
                    id={`popover-open-${id}`}
                    open={open}
                    anchorEl={anchorEl}
                    onClose={handleClose}
                    anchorOrigin={{
                        vertical: 'top',
                        horizontal: 'left'
                    }}
                    transformOrigin={{
                        vertical: 'bottom',
                        horizontal: 'left'
                    }}>
                    <Box padding={1}>
                        <Grid container spacing={1} alignContent="center" alignItems="center">
                            <Grid item>
                                <TextField
                                    fullWidth
                                    label="Tipo de campo"
                                    select
                                    size="small"
                                    value={type}
                                    name="type"
                                    onChange={handleChangeProps}
                                    className={classes.option}>
                                    {[TITLE, SUBTITLE, PARAGRAPH, QUOTES, CUSTOM_TEXT].map(
                                        ($type, $key) => (
                                            <MenuItem
                                                disabled={$type === type}
                                                value={$type}
                                                key={$key}>
                                                {TEXT_TYPES[$type]}
                                            </MenuItem>
                                        )
                                    )}
                                </TextField>
                            </Grid>
                            <Grid item>
                                <TextField
                                    fullWidth
                                    label="Alinear Texto"
                                    select
                                    size="small"
                                    value={align}
                                    defaultValue="start"
                                    name="align"
                                    onChange={handleChangeProps}
                                    className={classes.option}>
                                    <MenuItem value="start">Inicio</MenuItem>
                                    <MenuItem value="center">Centro</MenuItem>
                                    <MenuItem value="end">Fin</MenuItem>
                                </TextField>
                            </Grid>
                            <Grid item>
                                <TextField
                                    fullWidth
                                    label="Tamaño de Fuente"
                                    type="number"
                                    size="small"
                                    value={fontSize}
                                    disabled={type !== CUSTOM_TEXT}
                                    name="fontSize"
                                    onChange={handleChangeProps}
                                    className={classes.option}
                                />
                            </Grid>
                            <Grid item>
                                <TextField
                                    fullWidth
                                    label="Tamaño del objeto"
                                    select
                                    size="small"
                                    value={size}
                                    name="size"
                                    disabled={disableGrid}
                                    onChange={handleChangeProps}
                                    className={classes.option}>
                                    {GRID_SIZES.map(($size) => (
                                        <MenuItem
                                            disabled={size === $size.value}
                                            value={$size.value}
                                            key={`select-size-option-${id}-${$size.value}`}>
                                            {$size.label}
                                        </MenuItem>
                                    ))}
                                </TextField>
                            </Grid>
                            <Grid item>
                                <Tooltip title="Borrar" interactive>
                                    <ColorPick
                                        color={color}
                                        icon={FormatColorText}
                                        onChange={(rgb) => onChange({ color: rgb })}
                                    />
                                </Tooltip>
                            </Grid>
                            <Tooltip title="Insertar Emoji">
                                <EmojiPicker
                                    id={`emoji-picker-text-${id}`}
                                    onSelect={handlePushEmoji}
                                />
                            </Tooltip>
                        </Grid>
                        <Grid item>
                            <Grid item>
                                <Tooltip title="Borrar" interactive>
                                    <IconButton size="small" onClick={() => onDelete(id)}>
                                        <DeleteForever color="error" />
                                    </IconButton>
                                </Tooltip>
                            </Grid>
                        </Grid>
                    </Box>
                </Popover>
            )}
        </React.Fragment>
    );
};

TextInput.propTypes = {
    id: PropTypes.string.isRequired,
    value: PropTypes.string,
    color: PropTypes.string,
    type: PropTypes.string,
    fontSize: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    align: PropTypes.string,
    size: PropTypes.any,
    child: PropTypes.bool,
    disableGrid: PropTypes.bool,
    onChange: PropTypes.func.isRequired,
    onDelete: PropTypes.func.isRequired
};

TextField.defaultProps = {
    value: '',
    type: PARAGRAPH,
    color: '#000000',
    fontSize: 14,
    align: 'start',
    size: 12,
    child: false,
    disableGrid: false
};

export default TextInput;
