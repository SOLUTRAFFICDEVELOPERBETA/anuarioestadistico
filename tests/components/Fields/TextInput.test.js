import '@testing-library/jest-dom';
import { act } from '@testing-library/react';
import { mount } from 'enzyme';
import React from 'react';
import { DndProvider } from 'react-dnd';
import { HTML5Backend } from 'react-dnd-html5-backend';
import TextInput from '../../../components/Fields/TextInput';
import AlertContext from '../../../contexts/alert';

describe('Pruebas en <TextInput/>', () => {
    const onChange = jest.fn(() => { });
    const onDelete = jest.fn();
    const showMessage = jest.fn();
    const wrapper = mount(
        <DndProvider backend={HTML5Backend}>
            <AlertContext.Provider value={{
                showMessage
            }}>
                <TextInput
                    id={`id-${1}-Text-input`}
                    value={"contendido de prueba"}
                    type={"paragraph"}
                    size={12}
                    align='start'
                    fontSize={14}
                    child={false}
                    onChange={onChange}
                    onDelete={onDelete}
                />
            </AlertContext.Provider>
        </DndProvider>
    )
    test('debe mostrarse correctamente', () => {
       
        expect(wrapper).toMatchSnapshot();
    })

    test('debe llamar la funciÃ³n onChange()', () => {
        
        wrapper.find('span').at(1).prop('onBlur')({
            currentTarget: {
                innerText: 'innerText prueba de texto',
                localName: "span"
            }
        })

        expect(onChange).toHaveBeenCalled();
        expect(onChange).toHaveBeenCalledTimes(1);

    })


    test('debe mostrar el popover', () => {
        
        act(() => {
            wrapper.find('div').prop('onContextMenu')({
                preventDefault() { }, currentTarget: {
                    innerText: 'innerText de prueba nuevamente'
                }
            });
        })
        expect(wrapper).toMatchSnapshot();
       

    })
})