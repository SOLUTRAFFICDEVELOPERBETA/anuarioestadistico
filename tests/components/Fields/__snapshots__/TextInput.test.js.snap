// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TextInput/> debe mostrar el popover 1`] = `
Array [
  "",
  <div
    className=" css-1oyt5bq-InputContainer e11fp7a60"
    fontSize={14}
    id="id-1-Text-input"
    onContextMenu={[Function]}
  >
    <span
      className="dragger"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        style={
          Object {
            "fill": "#9e9e9e",
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M20 9H4v2h16V9zM4 15h16v-2H4v2z"
        />
      </svg>
    </span>
    <span
      className="text-field"
      contentEditable={true}
      dangerouslySetInnerHTML={
        Object {
          "__html": "contendido de prueba",
        }
      }
      data-type="paragraph"
      onBlur={[Function]}
    />
  </div>,
]
`;

exports[`Pruebas en <TextInput/> debe mostrarse correctamente 1`] = `
Array [
  "",
  <div
    className=" css-1oyt5bq-InputContainer e11fp7a60"
    fontSize={14}
    id="id-1-Text-input"
    onContextMenu={[Function]}
  >
    <span
      className="dragger"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root"
        focusable="false"
        style={
          Object {
            "fill": "#9e9e9e",
          }
        }
        viewBox="0 0 24 24"
      >
        <path
          d="M20 9H4v2h16V9zM4 15h16v-2H4v2z"
        />
      </svg>
    </span>
    <span
      className="text-field"
      contentEditable={true}
      dangerouslySetInnerHTML={
        Object {
          "__html": "contendido de prueba",
        }
      }
      data-type="paragraph"
      onBlur={[Function]}
    />
  </div>,
]
`;
